File: frontend/src/app/features/rankings/services/ranking.service.ts
import { Injectable } from "@angular/core";
import { environment } from "../../../../environments/environment";
import { HttpClient } from "@angular/common/http";
import { PaginatedResults, SearchParams } from "../../../shared/models/Search";
import { Observable } from "rxjs";
import { CreateRankingDto, RankingDataDto, RankingSearchDto, UpdateRankingDto } from "../models/Ranking";

@Injectable({
    providedIn: "root"
})
export class RankingService {
    private apiUrl: string = `${environment.apiUrl}/rankings`;

    constructor(
        private readonly http: HttpClient,
    ) {}

    searchRankings(userId: number, searchParams: SearchParams): Observable<PaginatedResults<RankingSearchDto>> {
        return this.http.get<PaginatedResults<RankingSearchDto>>(`${this.apiUrl}/search/user/${userId}`, { params: { ...searchParams } });
    }

    getRankingById(id: number): Observable<RankingDataDto> {
        return this.http.get<RankingDataDto>(`${this.apiUrl}/${id}`);
    }

    getWatchedDatesByUserId(userId: number): Observable<Date[]> {
        return this.http.get<Date[]>(`${this.apiUrl}/watched-dates/user/${userId}`);
    }

    isRankingTitleUnique(userId: number, title: string): Observable<boolean> {
        return this.http.get<boolean>(`${this.apiUrl}/ranking-title/${title}/user/${userId}`);
    }

    createRanking(rankingDto: CreateRankingDto): Observable<RankingDataDto> {
        return this.http.post<RankingDataDto>(this.apiUrl, rankingDto);
    }

    createRankings(rankingDtos: CreateRankingDto[]): Observable<RankingDataDto[]> {
        return this.http.post<RankingDataDto[]>(`${this.apiUrl}/bulk`, rankingDtos);
    }

    updateRanking(rankingDto: UpdateRankingDto): Observable<RankingDataDto> {
        return this.http.put<RankingDataDto>(this.apiUrl, rankingDto);
    }

    deleteRanking(id: number): Observable<void> {
        return this.http.delete<void>(`${this.apiUrl}/${id}`);
    }
    
    deleteRankings(ids: number[]): Observable<void> {
        return this.http.delete<void>(`${this.apiUrl}/bulk`, { body: ids });
    }
}

File: backend/src/main/java/com/moviemator/features/ranking/service/RankingService.java
package com.moviemator.features.ranking.service;

import com.moviemator.features.ranking.dto.CreateRankingDto;
import com.moviemator.features.ranking.dto.RankingDataDto;
import com.moviemator.features.ranking.dto.RankingSearchDto;
import com.moviemator.features.ranking.dto.UpdateRankingDto;
import com.moviemator.shared.search.models.PaginatedResults;
import com.moviemator.shared.search.models.SearchParams;

import java.util.List;

public interface RankingService {

    PaginatedResults<RankingSearchDto> searchRankings(Long userId, SearchParams searchParams);
    RankingDataDto getRankingById(Long id);
    boolean isRankingTitleUnique(Long userId, String title);
    RankingDataDto createRanking(CreateRankingDto rankingDto);
    List<RankingDataDto> createRankingsBulk(List<CreateRankingDto> rankingDtos);
    RankingDataDto updateRanking(UpdateRankingDto rankingDto);
    void deleteRanking(Long id);
    void deleteRankings(List<Long> ids);
}
